src/options.ts

```diff
@@ -0,0 +1,125 @@
+
+/**
+ * Error thrown when an invalid option is provided.
+ */
+export class OptionsError extends Error {}
+
+/**
+ * Helper class for parsing an configurable option from provided CLI flags
+ * or environment variables.
+ */
+class ConfigurableOption<T> {
+  constructor(
+    public readonly cliOption: string,
+    private envVar: string,
+    private defaultValue: T,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private validator: (x: any) => T = x => x as T
+  ) {}
+
+  parse(cliArgs: minimist.ParsedArgs, envVars: NodeJS.ProcessEnv): T {
+    return this.validator(
+      cliArgs[this.cliOption] || envVars[this.envVar] || this.defaultValue
+    );
+  }
+}
+
+const PortOption = new ConfigurableOption('port', 'PORT', '8080');
+const FunctionTargetOption = new ConfigurableOption(
+  'target',
+  'FUNCTION_TARGET',
+  'function'
+);
+const SourceLocationOption = new ConfigurableOption(
+  'source',
+  'FUNCTION_SOURCE',
+  '',
+  resolve
+);
+const SignatureOption = new ConfigurableOption(
+  'signature-type',
+  'FUNCTION_SIGNATURE_TYPE',
+  'http' as SignatureType,
+  x => {
+    if (isValidSignatureType(x)) {
+      return x;
+    }
+    throw new OptionsError(
+      `Function signature type must be one of: ${SignatureType.join(', ')}.`
+    );
+  }
+);
+
+export interface FrameworkOptions {
+  /**
+   * The port on which this server listens to all HTTP requests.
+   */
+  port: string;
+  /**
+   * The name of the function within user's node module to execute. If such a
+   * function is not defined, then falls back to 'function' name.
+   */
+  target: string;
+  /**
+   * The path to the source code file containing the client function.
+   */
+  sourceLocation: string;
+  /**
+   * The signature type of the client function.
+   */
+  signatureType: SignatureType;
+  /**
+   * Whether or not the --help CLI flag was provided.
+   */
+  printHelp: boolean;
+}
+
+export const helpText = `Example usage:
+functions-framework --target=helloWorld --port=8080
+Documentation:
+https://github.com/GoogleCloudPlatform/functions-framework-nodejs`;
+
+/**
+ * Parse the configurable framework options from the provided CLI arguments and
+ * environment variables.
+ * @param cliArgs the raw command line arguments
+ * @param envVars the environment variables to parse options from
+ * @returns the parsed options that should be used to configure the framework.
+ */
+export const parseOptions = (
+  cliArgs: string[] = process.argv,
+  envVars: NodeJS.ProcessEnv = process.env
+): FrameworkOptions => {
+  const argv = minimist(cliArgs, {
+    string: [
+      PortOption.cliOption,
+      FunctionTargetOption.cliOption,
+      SignatureOption.cliOption,
+      SourceLocationOption.cliOption,
+    ],
+  });
+  return {
+    port: PortOption.parse(argv, envVars),
+    target: FunctionTargetOption.parse(argv, envVars),
+    sourceLocation: SourceLocationOption.parse(argv, envVars),
+    signatureType: SignatureOption.parse(argv, envVars),
```

Comment:
Per comment above, this seems like it could be simplified as we're just adding a list of `ConfigurableOption`. Doesn't seem to pass the squint test.

Reply:
I spent an hour trying to change this, but I couldn't get it to come out any cleaner. I either ended up with a bunch more code, or losing the strong typing offered by the `FrameworkOptions` interface. Could we land this as it is and try and take another pass at cleanup in subsequent CL?

Reply:
OK np.